orttraining/orttraining/python/training/ortmodule/_custom_op_symbolic_registry.py

```diff
@@ -116,3 +119,252 @@ def avg_pool2d(g, self, kernel_size, stride, padding, ceil_mode, count_include_p
+    for label in input_labels:
+        # Lookup output index for label.
+        perm[label_perm_map[label]] = idx
+        idx += 1
+
+    # Add dimensions for missing labels.
+    for i in range(output_len):
+        if perm[i] == -1:
+            unsqueeze_axes.append(idx)
+            perm[i] = idx
+            idx += 1
+
+    return perm, unsqueeze_axes
+
+def unsqueeze_and_permute_for_mul(g, tensor, unsqueeze_axes, perm):
+    # If perm is sorted after removing unsqueeze axes, then permute is not needed.
+    # For example, a.unsqueeze(2).permute([0, 2, 1]) is same as a.unsqueeze(1).
+    if unsqueeze_axes:
+        new_perm = [v for v in perm if v not in unsqueeze_axes]
+        sorted = all(new_perm[i] < new_perm[i + 1] for i in range(len(new_perm) - 1))
+        if sorted:
+            return sym_help._unsqueeze_helper(g, tensor, [perm.index(axis) for axis in unsqueeze_axes])
+
+    if len(unsqueeze_axes) > 0:
+        tensor = sym_help._unsqueeze_helper(g, tensor, unsqueeze_axes)
+    if need_permute(perm):
+        tensor = g.op("Transpose", tensor, perm_i=perm)
+    return tensor
+
+def unsqueeze_and_permute_for_matmul(g, tensor, unsqueeze_axes, perm1, perm2):
+    # When going here, the unsqueeze axes must be some axes at the end.
+    # We can combine two permutes and remove unsqueeze axes, because we will reshape it after this.
+    # For example, a.unsqueeze([2,3]).permute([2,3,1,0]).permute([0,1,3,2])
+    # = a.unsqueeze([2,3]).permute([2,3,0,1]) = a.permute([0,1]) = a.
+    new_perm = [perm1[axis] for axis in perm2]
+    new_perm = [axis for axis in new_perm if axis not in unsqueeze_axes]
+    if need_permute(new_perm):
+        tensor = g.op("Transpose", tensor, perm_i=new_perm)
+    return tensor
+
+def shape_tensor_by_axes(g, input, input_shape, axes, need_reduce):
+    if input_shape is None:
+        input_shape = g.op("Shape", input)
+    shape_tensor = g.op("Gather", input_shape, g.op("Constant", value_t=torch.tensor(axes, dtype=torch.int64)), axis_i=0)
+    numel_tensor = None
+    if need_reduce:
+        numel_tensor = g.op("ReduceProd", shape_tensor) if len(axes) > 1 else shape_tensor
+    return shape_tensor, numel_tensor, input_shape
+
+def reshape_tensor(g, input, shape_tensors):
+    tensors = [tensor for tensor in shape_tensors if tensor is not None]
+    assert tensors
+    new_shape_tensor = g.op("Concat", *tensors, axis_i=0) if len(tensors) > 1 else tensors[0]
+    return g.op("Reshape", input, new_shape_tensor)
+
+@register_symbolic('einsum')
+@parse_args('s', 'v')
+def einsum(g, equation, tensor_list):
+    tensors = sym_help._unpack_list(tensor_list)
+    num_ops = len(tensors)
+    assert num_ops > 0
+
+    # Doesn't support implicit output is ellipsis or more than 2 oprands for now.
+    # Doesn't support ellipsis ('...') for now as not easy to get sizes of oprands.
+    if num_ops != 2 or equation.find('->') == -1 or '.' in equation:
+        return g.op("Einsum", *tensors, equation_s=equation)
```

Comment:
Just wondering why we don't implement this as an optimization in ORT (as a graph-rewrite for einsum), which would make it more general? Just trying to understand the reasons, not suggesting doing it now.

Reply:
We have the EinSum Op and kernel implemented in ORT for inference, but we don't have the gradient of it for training. This rewrite is needed for training only, and using Python code to implement this here is much easier than using C++ in ORT.